pkg/controller/factory/base_controller.go

```diff
@@ -0,0 +1,158 @@
+
+	for i := 1; i <= workers; i++ {
+		klog.Infof("Starting #%d worker of %s controller ...", i, c.name)
+		workerWaitGroup.Add(1)
+		go func() {
+			defer func() {
+				klog.Infof("Shutting down worker of %s controller ...", c.name)
+				workerWaitGroup.Done()
+			}()
+			c.runWorker(queueContext)
+		}()
+	}
+
+	// runPeriodicalResync is independent from queue:w
+	go c.runPeriodicalResync(ctx, c.resyncEvery)
+
+	// Handle controller shutdown
+
+	<-ctx.Done() // wait for controller context to be cancelled
+	klog.Infof("Shutting down %s ...", c.name)
+
+	c.syncContext.Queue().ShutDown() // shutdown the controller queue first
+	for !c.syncContext.Queue().ShuttingDown() {
+		time.Sleep(10 * time.Millisecond) // don't race too much with the queue
+	}
+	queueShutdownComplete() // signal workers the queue is terminated
+
+	// wait for all workers to finish their jobs
+	workerWaitGroup.Wait()
+
+	shutdownComplete() // cancel the shutdown context
+}
+
+func (c *baseController) Sync(ctx context.Context, syncCtx SyncContext) error {
+	return c.sync(ctx, syncCtx)
+}
+
+func (c *baseController) ShutdownContext() context.Context {
+	return c.shutdownContext
+}
+
+func (c *baseController) runPeriodicalResync(ctx context.Context, interval time.Duration) {
+	if interval == 0 {
+		return
+	}
+	go wait.UntilWithContext(ctx, func(ctx context.Context) {
+		if err := c.sync(ctx, c.syncContext.(syncContext).withRuntimeObject(nil)); err != nil {
+			utilruntime.HandleError(fmt.Errorf("periodical resync of controller %s failed: %v", c.name, err))
+		}
+	}, interval)
+}
+
+// runWorker runs a single worker
+// The worker is asked to terminate when the passed context is cancelled and is given terminationGraceDuration time
+// to complete its shutdown.
+func (c *baseController) runWorker(queueCtx context.Context) {
+	// workerContext is used to signal when the processing of the current item is done and the queue is shutdown
+	workerContext, workerDone := context.WithCancel(context.Background())
+
+	go func() {
+		defer workerDone()
+		for {
+			select {
+			case <-queueCtx.Done():
+				return
+			default:
+				c.processNextWorkItem(queueCtx)
+			}
+		}
+	}()
+
+	// wait for the queue to be marked as shutdown
+	<-queueCtx.Done()
+
+	// sets a counter for the processor to finish its job.
+	// by using the workerContext as parent context either the workerContext is complete or terminationGraceDuration pass (whatever comes first).
+	// this allows the worker to gracefully terminate its job.
+	workerDeadline, workerDeadlineCancel := context.WithTimeout(workerContext, c.terminationGraceDuration)
+	defer workerDeadlineCancel() // this is not really needed, but govet complains about context cancel leak...
+	<-workerDeadline.Done()
+}
+
+func (c *baseController) processNextWorkItem(queueCtx context.Context) {
+	syncObject, quit := c.syncContext.Queue().Get()
+	if quit {
+		return
+	}
+
+	defer c.syncContext.Queue().Done(syncObject)
+
+	runtimeObj, ok := syncObject.(runtime.Object)
```

Comment:
if you have multiple informers, it will make it harder to determine what GVK the key is... this works, but we don't use this much anyway. I don't think having this option will hurt us.

Reply:
unless @deads2k or @sttts know a reason why using an object as a queue key is really bad idea.